### ISSUE TITLE
Publish "Annex of Appendix of Brochure" documents

### ISSUE THREAD

**Issue by opoudjis** [metanorma/metanorma-bipm#99]
Follow on from https://github.com/metanorma/metanorma-bipm/issues/98, which is a follow on from https://github.com/metanorma/bipm-si-brochure/issues/71

Not only are we publishing Appendixes to the BIPM Brochure separately, we are now also publishing Annexes to Appendixes to the BIPM Brochure separately.

This means that you will need not only `//bibdata/title[@type = 'appendix']` and `//bibdata/extstructuredidentifier/appendix`, but also `//bibdata/title[@type = 'annex']` and `//bibdata/extstructuredidentifier/annexid`. (Note: annexid, not annex, to avoid Xpath ambiguity.)

You will need to ask @ronaldtse where he wants those displayed on the document.

**Comment by Intelligent2013** [comment]
To do:
1. hierarchy/order to show titles on cover page in PDF (title/@type):
  - main/cover 
    - appendix (number from structuredidentifier/_appendix_)
      - annex (number from structuredidentifier/_annexid_)
        - part (number from structuredidentifier/_part_)
            - subpart (number from structuredidentifier/_subpart_)
           
2. Adaptive font-size depends on text length.

**Comment by Intelligent2013** [comment]
@ronaldtse 
Before each pair of titles we display number with type-of-title prefix, first for French, second for English. Example for Appendix:
![изображение](https://user-images.githubusercontent.com/5559926/103551851-06cf5000-4ebc-11eb-8023-ae13a9bd22ca.png)

What is title prefix should we show for 'Annex' (i,e. 'Annex of Appendix') titles?
If we'll show:

Annexe N (in French)
**Annex N** (in English)

then in this case there is a confusion between titles (Appendix vs. Annex) - Annexe in both cases for French.

**Comment by Intelligent2013** [comment]
@ronaldtse regarding your notices:
```
The SI Brochure contains Appendix 2.
Appendix 2 is the "Mise en pratique".
Appendix 2 contains multiple "part documents". Each document is encoded to have a "part number" and a "part title".
A "part" inside Appendix 2 can contain Annexes (and the Annexes are published separately). Each Annex can have an "annex ID" and "annex title".

What we need to do on the cover:

1. The "part number" and "part title" is not displayed on the cover.
2. The "annex id" and "annex title" is not displayed on the cover.
3. When the "appendix name" is long, the cover page overflows...
```
I've updated bipm xslt to 
- adaptive font-size depends of length of titles (but I would like to try improve the algorithm...)
- part number and part title display now,
- To display annex id and annex title, you need to add title with type `annex`, I've added `title/@type=annex` processing.
- To display Appendix 2 title "Mise en pratique" you need to add title with type `appendix`.

So my manually updated source xml  looks as:
```xml
<title language="en" format="text/plain" type="main">The International System of Units</title>
<title language="en" format="text/plain" type="appendix">Mise en pratique</title>
<title language="en" format="text/plain" type="annex">Recommended values of standard frequencies for applications including the practical realization of the metre and secondary representations of the second</title>
<title language="en" format="text/plain" type="part">Rubidium 87 Atom (<stem type="MathML"><math xmlns="http://www.w3.org/1998/Math/MathML"><mi>f</mi><mo>≈</mo><mn>6.835</mn><mtext> </mtext><mstyle mathvariant="normal"><mrow><mi>G</mi><mi>H</mi><mi>z</mi></mrow></mstyle></math></stem>)</title>
<title language="fr" format="text/plain" type="main">Le système international d’unités</title>
<title language="fr" format="text/plain" type="appendix">Mise en pratique</title>
<title language="fr" format="text/plain" type="annex">Valeurs recommandées des fréquences étalons destinées à la mise en pratique de la définition du mètre et aux représentations secondaires de la seconde</title>
<title language="fr" format="text/plain" type="part">Rubidium 87 Atom (<stem type="MathML"><math xmlns="http://www.w3.org/1998/Math/MathML"><mi>f</mi><mo>≈</mo><mn>6.835</mn><mtext> </mtext><mstyle mathvariant="normal"><mrow><mi>G</mi><mi>H</mi><mi>z</mi></mrow></mstyle></math></stem>)</title>
...
<ext>
...
<part>2.29</part>
<appendix>2</appendix>
<annexid>3</annexid>
</structuredidentifier>
</ext>
```
and resulted PDF cover page:
![изображение](https://user-images.githubusercontent.com/5559926/103560627-215bf600-4ec9-11eb-8247-649e898abf9f.png)

Resulted PDF: 
[mep-wgfs-cm4_4-Rb_6-835GHz.presentation.pdf](https://github.com/metanorma/metanorma-bipm/files/5765924/mep-wgfs-cm4_4-Rb_6-835GHz.presentation.pdf)

**Comment by ronaldtse** [comment]
@Intelligent2013 the current layout is acceptable, will consult BIPM on the French of "Annex" vs "Appendix". Thanks!

**Comment by opoudjis** [comment]
From memory, they have flipped them around: appendice for annex, annexe for appendix

**Comment by opoudjis** [comment]
No, I am still on holiday :p I think the xml i am generating is at odds with some of this, but we will find out in a couple of days

**Comment by ronaldtse** [comment]
@Intelligent2013 1 thing to fix here: "May , 2016" has a space in between.

I have consulted BIPM on the difference between Annex/Appendix in French/English and will post in a new issue.

**Comment by ronaldtse** [comment]
Can we make this a configurable option @opoudjis ? e.g. `:annex-name-fr: Annexe {{ number }}`, or just ":bipm-old-hierarchy:"?

**Comment by opoudjis** [comment]
Pending clarification: BIPM seem to be applying document autonomy, not hierarchical position, as their naming criterion.

**Comment by ronaldtse** [comment]
I have requested further clarification from BIPM:

> I’d like to confirm the roles of “Appendix” and “Annex” (in English as representation of the concepts) being the “types of major sections”.
> 
Should we use a hierarchical system in determining the type of a major section? Or is the type of major section dependent on the amount of content inside it?
> 
> Option 1:
> Using a defined hierarchical system, such as that used by ISO/IEC and other SDOs like ITU.
> In ISO they use Level 1 (major section: “Document") > Level 2 (“Annex") > Level 3 (“Appendix"). 
> 
> This system uses a strict hierarchy:
> - only a Document is allowed to have an Annex
> - only an Annex is allowed to have an Appendix
> 
> Option 2:
> Using a system where the amount of content (defined by the author) determines the type of major section:
> - Any document can contain both Annexes and Appendices
> - An Annex can contain an Appendix
> 
> This will raise additional questions such as: 
> - Can an Appendix can contain an Annex?
> - Can an Annex contain an Annex?
> - Can an Appendix contain an Appendix?

**Comment by Intelligent2013** [comment]
@ronaldtse 
> 1 thing to fix here: "May , 2016" has a space in between.

Fixed. Comma was related to day, but date is month+year only.

> For now, we will use:
> Level 1 — Brochure
> Level 2 — Annexe 2 (en: Appendix 2)
> Level 3 — Annexe 1 (en: Annex X)

> So it will be "Annexe of Annexe" in French for now.

So no need to change xslt at this moment. Thank you.

**Comment by ronaldtse** [comment]
Thanks!

**Comment by opoudjis** [comment]
Update on the Options query needed, @ronaldtse

**Comment by ronaldtse** [comment]
@opoudjis we are still waiting for a discussion to happen in the BIPM, no updates on this yet. Completing the main SI Brochure first will bring that discussion forward...

**Comment by ronaldtse** [comment]
BIPM lead for machine readable Janet Miles has suggested that this change may be done in the next edition (10) of the official SI Brochure. So this is not going to be done right now, but perhaps a few months down the line.

**Comment by opoudjis** [comment]
Enforce consistent hierarchy unless told otherwise, and assume this for edns 9 and 10.

**Comment by opoudjis** [comment]
This is confusing, and has been unactioned too long for it to make sense any more, but let me restate what I think is going on here.

The request appears to be:

* Editions 9 and 10 will impose a consistent hierarchy, Brochure > Annex > Appendix
* Edition 8 and earlier appear to impose a hierarchy Brochure > Appendix > Annex in English, Brochure > Annexe > Annexe in French (!)
* Historically, the choice of what to call a Level 2 artefact, appendice or annexe, depended on its extent, and was subjective
* If possible, do not change the metadata

The legacy option of alternative between appendice and annexe depending on extent has not actually been requested by BIPM, and I'm not going to implement it unless requested. So right now, the hierarchy is fixed, and BIPM has accepted that the hierarchy is fixed; the only change is in the labels assigned Level 2 and 3.

The sample given in https://github.com/metanorma/metanorma-bipm/issues/98

````
= MeP-K Low-temperature Johnson Noise Thermometry
:partnumber: 1
:edition: 9
:copyright-year: 2019
:revdate: 2019-05-20
:language: en
:title-annex-en: MeP-K Low-temperature Johnson Noise Thermometry         <<< current document title
:title-annex-fr: MeP-KThermométrie de bruit Johnson à basse température <<< current document title
:title-appendix-en: Mise en pratique for the definition of the kelvin in the SI <<< parent document title
:title-appendix-fr: Mise en pratique de la définition du kelvin <<< parent document title
:title-en: The International System of Units <<< grandparent document title
:title-fr: Le système international d’unités <<< grandparent document title
:appendix-id: 2         <<< parent document locator in grandparent document
:annex-id: 3        <<< current document id in the annex
:doctype: mise-en-pratique
:parent-document: mep-kelvin-definition-en.adoc
:docnumber: SI MEP K LT
:committee-acronym: CCT
:committee-en: Consultative Committee for Thermometry
:committee-fr: Comité consultatif de thermométrie
:si-aspect: K_k
````

has the current completely inconsistent rendering, which is a testament to what a clown show this has been.

* Document identifier, English: Appendix > Annex
* Document identifier, French: Annexe > Appendice
* HTML Title labels, English: Appendix > Annex
* HTML Title labels, French: Annexe > Appendice

* PDF Title labels, English: Appendix > Annex
* PDF Title labels, French: Annexe > Annexe

PDF has implemented what as asked for for edition 8:

> For now, we will use:
Level 1 — Brochure
Level 2 — Annexe 2 (en: Appendix 2)
Level 3 — Annexe 1 (en: Annex X)

<img width="645" alt="Image" src="https://github.com/user-attachments/assets/0e3e6f05-922d-43bc-92df-a17bd97d2fe3" />

HTML and document identifiers (they draw on the same implementation) has implemented neither the old nor the new request: it's correct for English, but the other way around for French.

````
<docidentifier primary="true" type="BIPM" language="en">BIPM SI MEP K LT Appendix 2 Annex 3 Part 1</docidentifier>
<docidentifier primary="true" type="BIPM" language="fr">BIPM SI MEP K LT Annexe 2 Appendice 3 Partie 1</docidentifier>
````

<img width="945" alt="Image" src="https://github.com/user-attachments/assets/8d5ce476-b52d-4ae5-b0e9-3fe0ab15fce6" />


The new implementation is going to flip again.

So:

Current HTML: Appendix > Annex, Annexe > Appendice
Current PDF: Appendix > Annex, Annexe > Annexe
Request, Edition 8: Appendix > Annex, Annexe > Annexe
Request, Edition 9/10: Annex > Appendix, Annexe > Appendice

The i18n files currently look like:

````
annex: Appendix
appendix: Annex
iso_annex: Annex
````

````
annex: Annexe
appendix: Appendice
iso_annex: Annexe
````

In order to prevent future confusion, I am going to change the i18n files to make them sematically labelled: not `annex` and `appendix` as the label identifiers, but `level2_ancillary` and `level3_ancillary`. And the value of level2_ancillary and level3_ancillary is going to depend on :`document-scheme:`:

If document_scheme >= 2021 (@ronaldtse please confirm cutoff):

````
level2_ancillary: Annex
level3_ancillary: Appendix

level2_ancillary: Annexe
level3_ancillary: Appendice
````

If document_scheme < 2021 (current PDF):

````
level2_ancillary: Appendix
level3_ancillary: Annex

level2_ancillary: Annexe
level3_ancillary: Annexe
````


Both HTML (lib/isodoc/bipm/metadata.rb) and PDF will need to read their values for what to label annexes and appendices from `level2_ancillary`, `level3_ancillary`, and ignore `annex`, `appendix` as i18n labels.

@ronaldtse Please confirm you're ok with this. If a legacy document has idiosyncratic labelling of Level 2 or 3, we can always supply a local i18n file to override `level2_ancillary`, `level3_ancillary` during compilation.

**Comment by ronaldtse** [comment]
Agree. We should allow the ancillary locality label customizable through i18n, and allow document attribute override of that per document.

**Comment by Intelligent2013** [comment]
@opoudjis I've updated BIPM XSLT to render values from `level2_ancillary`, `level2_ancillary_alt`, `level3_ancillary` and `level3_ancillary_alt` (instead of hard-coded constants in the XSLT).

PDF examples:
- `:document-scheme: 2021`:
![Image](https://github.com/user-attachments/assets/b8184b84-71ef-4fd0-bafe-c05d5bb085e0)

- `:document-scheme: 2019`:
![Image](https://github.com/user-attachments/assets/8edf23f2-52c2-4166-a761-8f4d1134d37a)

In the BIPM still there are the hard-coded labels for `:title-part-...:`:
- `Part` (en)
- `Partie` (fr)

and `:title-subpart-...:`:
- `Sub-part` (en)
- `Partie de sub` (fr)

Could you add the i18n variables `level4_ancillary`, `level5_ancillary`, `level4_ancillary_alt`, `level5_ancillary_alt` similarly for part/sub-parts? Thanks!

**Comment by opoudjis** [comment]
Refactoring to create the language-specific identifiers in Presentation XML, using the correct level labels.

**Comment by opoudjis** [comment]
@Intelligent2013, please try now. I've refactored the code, but you should not be impacted by the rearranged docidentifier, so long as you are using `docidentifier[@type = 'BIPM'][@language]`

**Comment by Intelligent2013** [comment]
@opoudjis thank you! BIPM XSLT updated.
